evm/src/arithmetic/byte.rs

```diff
@@ -0,0 +1,502 @@
+
+    // Store 256 * lo rather than lo:
+    lv[BYTE_LAST_LIMB_LO] = F::from_canonical_u64(lo << 8);
+    lv[BYTE_LAST_LIMB_HI] = F::from_canonical_u64(hi);
+
+    let tree = &mut lv[AUX_INPUT_REGISTER_1];
+    let output = if idx.bit(0) {
+        tree[15] = F::from_canonical_u64(lo);
+        lo.into()
+    } else {
+        tree[15] = F::from_canonical_u64(hi);
+        hi.into()
+    };
+
+    u256_to_array(
+        &mut lv[OUTPUT_REGISTER],
+        if idx < 32.into() {
+            output
+        } else {
+            U256::zero()
+        },
+    );
+}
+
+pub fn eval_packed<P: PackedField>(
+    lv: &[P; NUM_ARITH_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let is_byte = lv[IS_BYTE];
+
+    let idx = &lv[INPUT_REGISTER_0];
+    let val = &lv[INPUT_REGISTER_1];
+    let out = &lv[OUTPUT_REGISTER];
+    let idx_decomp = &lv[AUX_INPUT_REGISTER_0];
+    let tree = &lv[AUX_INPUT_REGISTER_1];
+
+    // low 5 bits of the first limb of idx:
+    let mut idx0_lo5 = P::ZEROS;
+    for i in 0..5 {
+        let bit = idx_decomp[i];
+        yield_constr.constraint(is_byte * (bit * bit - bit));
+        idx0_lo5 += bit * P::Scalar::from_canonical_u64(1 << i);
+    }
+    // high (11) bits of the first limb of idx:
+    let idx0_hi = idx_decomp[5] * P::Scalar::from_canonical_u64(32u64);
+    yield_constr.constraint(is_byte * (idx[0] - (idx0_lo5 + idx0_hi)));
+
+    // Verify the layers of the tree
+    // NB: Each of the bit values is negated in place to account for
+    // the reversed indexing.
+    let bit = idx_decomp[4];
+    for i in 0..8 {
+        let limb = bit * val[i] + (P::ONES - bit) * val[i + 8];
+        yield_constr.constraint(is_byte * (tree[i] - limb));
+    }
+
+    let bit = idx_decomp[3];
+    for i in 0..4 {
+        let limb = bit * tree[i] + (P::ONES - bit) * tree[i + 4];
+        yield_constr.constraint(is_byte * (tree[i + 8] - limb));
+    }
+
+    let bit = idx_decomp[2];
+    for i in 0..2 {
+        let limb = bit * tree[i + 8] + (P::ONES - bit) * tree[i + 10];
+        yield_constr.constraint(is_byte * (tree[i + 12] - limb));
+    }
+
+    let bit = idx_decomp[1];
+    let limb = bit * tree[12] + (P::ONES - bit) * tree[13];
+    yield_constr.constraint(is_byte * (tree[14] - limb));
+
+    // Check byte decomposition of last limb:
+
+    let base8 = P::Scalar::from_canonical_u64(1 << 8);
+    // base8_inv is Goldilocks-specific
+    debug_assert!(P::Scalar::order() == 0xffff_ffff_0000_0001u64.into());
+    let base8_inv = P::Scalar::from_canonical_u64(0xfeffffff01000001);
+    debug_assert!(base8 * base8_inv == P::Scalar::ONE);
+
+    let lo_byte = lv[BYTE_LAST_LIMB_LO];
+    let hi_byte = lv[BYTE_LAST_LIMB_HI];
+    yield_constr.constraint(is_byte * (base8_inv * lo_byte + base8 * hi_byte - limb));
+
+    let bit = idx_decomp[0];
+    let t = bit * base8_inv * lo_byte + (P::ONES - bit) * hi_byte;
+    yield_constr.constraint(is_byte * (tree[15] - t));
+    let expected_out_byte = tree[15];
```

Comment:
To make the constraints field-agnostic but still avoiding doing inverses, we can rewrite the constraints to remove `base8_inv`. Just multiply each constraint by `base8`.

Something like:
```suggestion
    let base8 = P::Scalar::from_canonical_u64(1 << 8);
    // base8_inv is Goldilocks-specific
    debug_assert!(P::Scalar::order() == 0xffff_ffff_0000_0001u64.into());
    let base8_inv = P::Scalar::from_canonical_u64(0xfeffffff01000001);
    debug_assert!(base8 * base8_inv == P::Scalar::ONE);

    let lo_byte = lv[BYTE_LAST_LIMB_LO];
    let hi_byte = lv[BYTE_LAST_LIMB_HI];
    yield_constr.constraint(is_byte * (lo_byte + base8 * (base8 * hi_byte - limb)));

    let bit = idx_decomp[0];
    let t = bit * lo_byte + (P::ONES - bit) * base8 * hi_byte;
    yield_constr.constraint(is_byte * (base8 * tree[15] - t));
    let expected_out_byte = tree[15];
```